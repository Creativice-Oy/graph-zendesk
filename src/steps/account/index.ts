import {
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';
import { createAPIClient } from '../../client';
import { IntegrationConfig } from '../../config';
import { createAccountEntity } from './converters';
import { ACCOUNT_ENTITY_KEY, Entities, IntegrationSteps } from '../constants';

// Oauth scope: 'read'
export async function fetchAccount({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  const currentUser = await apiClient.getCurrentUser();
  const accountEntity = createAccountEntity(currentUser);
  await jobState.addEntity(accountEntity);
  await jobState.setData(ACCOUNT_ENTITY_KEY, accountEntity);
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: IntegrationSteps.ACCOUNT,
    name: 'Fetch Account',
    entities: [Entities.ACCOUNT],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccount,
  },
];
