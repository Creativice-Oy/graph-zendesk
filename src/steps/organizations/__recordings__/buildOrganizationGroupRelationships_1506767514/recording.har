{
  "log": {
    "_recordingName": "buildOrganizationGroupRelationships",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "5.1.1"
    },
    "entries": [
      {
        "_id": "ed2fd5f5d5cd127ce3120f22e7ddec13",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "jupiterone-dev.zendesk.com"
            }
          ],
          "headersSize": 330,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "page",
              "value": "1"
            }
          ],
          "url": "https://jupiterone-dev.zendesk.com/api/v2/groups.json?page=1"
        },
        "response": {
          "bodySize": 426,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 426,
            "text": "{\"groups\":[{\"url\":\"https://jupiterone-dev.zendesk.com/api/v2/groups/360021738231.json\",\"id\":360021738231,\"name\":\"Support\",\"description\":\"\",\"default\":true,\"deleted\":false,\"created_at\":\"2021-10-10T20:49:30Z\",\"updated_at\":\"2021-10-10T20:49:30Z\"}],\"next_page\":null,\"previous_page\":null,\"count\":1}"
          },
          "cookies": [
            {
              "domain": ".jupiterone-dev.zendesk.com",
              "httpOnly": true,
              "name": "__cfruid",
              "path": "/",
              "sameSite": "None",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "date",
              "value": "Fri, 28 Jan 2022 14:53:11 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "zendesk-api-version",
              "value": "2022-01-01"
            },
            {
              "name": "x-zendesk-api-version",
              "value": "v2"
            },
            {
              "name": "x-zendesk-application-version",
              "value": "v10911"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "access-control-allow-origin",
              "value": "*"
            },
            {
              "name": "access-control-allow-credentials",
              "value": "true"
            },
            {
              "name": "access-control-expose-headers",
              "value": "X-Zendesk-API-Warn,X-Zendesk-User-Id,X-Zendesk-User-Session-Expires-At"
            },
            {
              "name": "x-rate-limit",
              "value": "200"
            },
            {
              "name": "x-rate-limit-remaining",
              "value": "188"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000;"
            },
            {
              "name": "etag",
              "value": "W/\"76e3a2fd91ec6717583c8b3cf6682ee8\""
            },
            {
              "name": "cache-control",
              "value": "max-age=0, private, must-revalidate"
            },
            {
              "name": "x-zendesk-origin-server",
              "value": "classic-app-server-7f77698bc-dhhzn"
            },
            {
              "name": "x-request-id",
              "value": "6d4b0c8258682c9b-SEA, 6d4b0c8258682c9b-SEA"
            },
            {
              "name": "x-runtime",
              "value": "0.157789"
            },
            {
              "name": "x-zendesk-zorg",
              "value": "yes"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "cf-cache-status",
              "value": "DYNAMIC"
            },
            {
              "name": "expect-ct",
              "value": "max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\""
            },
            {
              "name": "report-to",
              "value": "{\"endpoints\":[{\"url\":\"https:\\/\\/a.nel.cloudflare.com\\/report\\/v3?s=XPTIwuN92VUgnkukFKMtQDBHcQzyLBrZgV5XahDd4%2F%2FjMHDyHr6ig%2FNeYTX7%2B6p%2Bx%2FOWzfWVFSYGhfHGLym%2F69SYCq%2FAyYDdiZM3sIEO5OpGCDHQmMivZ%2FofPjthAr%2FZdrvxDABzdtrE3TdQ\"}],\"group\":\"cf-nel\",\"max_age\":604800}"
            },
            {
              "name": "nel",
              "value": "{\"success_fraction\":0.01,\"report_to\":\"cf-nel\",\"max_age\":604800}"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "server",
              "value": "cloudflare"
            },
            {
              "name": "cf-ray",
              "value": "6d4b0c8258682c9b-ORD"
            }
          ],
          "headersSize": 1523,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-01-28T14:53:11.248Z",
        "time": 330,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 330
        }
      },
      {
        "_id": "de45a5ef1f284a357f682f3933d50fc4",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "jupiterone-dev.zendesk.com"
            }
          ],
          "headersSize": 337,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "page",
              "value": "1"
            }
          ],
          "url": "https://jupiterone-dev.zendesk.com/api/v2/organizations.json?page=1"
        },
        "response": {
          "bodySize": 574,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 574,
            "text": "{\"organizations\":[{\"url\":\"https://jupiterone-dev.zendesk.com/api/v2/organizations/361584776731.json\",\"id\":361584776731,\"name\":\"JupiterOne\",\"shared_tickets\":false,\"shared_comments\":false,\"external_id\":null,\"created_at\":\"2021-10-10T20:49:30Z\",\"updated_at\":\"2021-11-05T16:50:42Z\",\"domain_names\":[],\"details\":\"\",\"notes\":\"\",\"group_id\":360021738231,\"tags\":[],\"organization_fields\":{}}],\"next_page\":null,\"previous_page\":null,\"count\":1}"
          },
          "cookies": [
            {
              "domain": ".jupiterone-dev.zendesk.com",
              "httpOnly": true,
              "name": "__cfruid",
              "path": "/",
              "sameSite": "None",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "date",
              "value": "Fri, 28 Jan 2022 14:53:12 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "zendesk-api-version",
              "value": "2022-01-01"
            },
            {
              "name": "x-zendesk-api-version",
              "value": "v2"
            },
            {
              "name": "x-zendesk-application-version",
              "value": "v10911"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "etag",
              "value": "W/\"aee7c5535e2c5f8d50b0f209ef410549\""
            },
            {
              "name": "access-control-allow-origin",
              "value": "*"
            },
            {
              "name": "access-control-allow-credentials",
              "value": "true"
            },
            {
              "name": "access-control-expose-headers",
              "value": "X-Zendesk-API-Warn,X-Zendesk-User-Id,X-Zendesk-User-Session-Expires-At"
            },
            {
              "name": "cache-control",
              "value": "max-age=30, private"
            },
            {
              "name": "x-rate-limit",
              "value": "200"
            },
            {
              "name": "x-rate-limit-remaining",
              "value": "186"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000;"
            },
            {
              "name": "x-zendesk-origin-server",
              "value": "classic-app-server-7f77698bc-h8qwx"
            },
            {
              "name": "x-request-id",
              "value": "6d4b0c847f38637e-SEA, 6d4b0c847f38637e-SEA"
            },
            {
              "name": "x-runtime",
              "value": "0.170101"
            },
            {
              "name": "x-zendesk-zorg",
              "value": "yes"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "cf-cache-status",
              "value": "DYNAMIC"
            },
            {
              "name": "expect-ct",
              "value": "max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\""
            },
            {
              "name": "report-to",
              "value": "{\"endpoints\":[{\"url\":\"https:\\/\\/a.nel.cloudflare.com\\/report\\/v3?s=cwSOtMOdmMFOmm0wOfCOVybxLpR2lnocQoKrW%2ByKdu4FpPu79HPRmApyv6wVvHKH9hg8fzNOY0HpbLD19c7oJt2ct%2F5yIjQx3OfZJ4p2ccwU8xX8Wb%2BPgrvBUpbB6fLPkrtsRPF9egYdEoRS\"}],\"group\":\"cf-nel\",\"max_age\":604800}"
            },
            {
              "name": "nel",
              "value": "{\"success_fraction\":0.01,\"report_to\":\"cf-nel\",\"max_age\":604800}"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "server",
              "value": "cloudflare"
            },
            {
              "name": "cf-ray",
              "value": "6d4b0c847f38637e-ORD"
            }
          ],
          "headersSize": 1493,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-01-28T14:53:11.581Z",
        "time": 365,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 365
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
