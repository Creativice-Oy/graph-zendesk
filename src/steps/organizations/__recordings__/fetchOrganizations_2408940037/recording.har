{
  "log": {
    "_recordingName": "fetchOrganizations",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "5.1.1"
    },
    "entries": [
      {
        "_id": "6963c6f3f589a3e055a06cafb0ba8004",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "jupiterone-dev.zendesk.com"
            }
          ],
          "headersSize": 319,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://jupiterone-dev.zendesk.com/api/v2/account"
        },
        "response": {
          "bodySize": 362,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 362,
            "text": "{\"account\":{\"url\":\"https://jupiterone-dev.zendesk.com\",\"name\":\"JupiterOne\",\"sandbox\":false,\"subdomain\":\"jupiterone-dev\",\"time_format\":24,\"time_zone\":\"America/New_York\",\"owner_id\":420755677092,\"multiproduct\":true}}"
          },
          "cookies": [
            {
              "domain": ".jupiterone-dev.zendesk.com",
              "httpOnly": true,
              "name": "__cfruid",
              "path": "/",
              "sameSite": "None",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "date",
              "value": "Fri, 28 Jan 2022 14:53:10 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "zendesk-api-version",
              "value": "2022-01-01"
            },
            {
              "name": "x-zendesk-api-version",
              "value": "v2"
            },
            {
              "name": "x-zendesk-application-version",
              "value": "v10911"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "access-control-allow-origin",
              "value": "*"
            },
            {
              "name": "access-control-allow-credentials",
              "value": "true"
            },
            {
              "name": "access-control-expose-headers",
              "value": "X-Zendesk-API-Warn,X-Zendesk-User-Id,X-Zendesk-User-Session-Expires-At"
            },
            {
              "name": "x-rate-limit",
              "value": "200"
            },
            {
              "name": "x-rate-limit-remaining",
              "value": "195"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000;"
            },
            {
              "name": "etag",
              "value": "W/\"a80e3e81a54a847723877a9cbb1366d9\""
            },
            {
              "name": "cache-control",
              "value": "max-age=0, private, must-revalidate"
            },
            {
              "name": "x-zendesk-origin-server",
              "value": "classic-app-server-7f77698bc-thtrq"
            },
            {
              "name": "x-request-id",
              "value": "6d4b0c7c1ce0c528-SEA, 6d4b0c7c1ce0c528-SEA"
            },
            {
              "name": "x-runtime",
              "value": "0.170958"
            },
            {
              "name": "x-zendesk-zorg",
              "value": "yes"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "cf-cache-status",
              "value": "DYNAMIC"
            },
            {
              "name": "expect-ct",
              "value": "max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\""
            },
            {
              "name": "report-to",
              "value": "{\"endpoints\":[{\"url\":\"https:\\/\\/a.nel.cloudflare.com\\/report\\/v3?s=IkIZkcH0aoFi3BZXfpjVI20mnX4sufVpCcrCSyWEgRYA4l79335%2Bsr89mdjBo1hK9uu%2BItfJLPBo%2FuaLArL48ybR%2Bw%2B%2FfPEJVNSuZffsomo4O5xJJrEPuU8SsUqBURR4RPIatTAwBBXtN5hR\"}],\"group\":\"cf-nel\",\"max_age\":604800}"
            },
            {
              "name": "nel",
              "value": "{\"success_fraction\":0.01,\"report_to\":\"cf-nel\",\"max_age\":604800}"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "server",
              "value": "cloudflare"
            },
            {
              "name": "cf-ray",
              "value": "6d4b0c7c1ce0c528-ORD"
            }
          ],
          "headersSize": 1515,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-01-28T14:53:10.162Z",
        "time": 438,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 438
        }
      },
      {
        "_id": "de45a5ef1f284a357f682f3933d50fc4",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "jupiterone-dev.zendesk.com"
            }
          ],
          "headersSize": 337,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "page",
              "value": "1"
            }
          ],
          "url": "https://jupiterone-dev.zendesk.com/api/v2/organizations.json?page=1"
        },
        "response": {
          "bodySize": 574,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 574,
            "text": "{\"organizations\":[{\"url\":\"https://jupiterone-dev.zendesk.com/api/v2/organizations/361584776731.json\",\"id\":361584776731,\"name\":\"JupiterOne\",\"shared_tickets\":false,\"shared_comments\":false,\"external_id\":null,\"created_at\":\"2021-10-10T20:49:30Z\",\"updated_at\":\"2021-11-05T16:50:42Z\",\"domain_names\":[],\"details\":\"\",\"notes\":\"\",\"group_id\":360021738231,\"tags\":[],\"organization_fields\":{}}],\"next_page\":null,\"previous_page\":null,\"count\":1}"
          },
          "cookies": [
            {
              "domain": ".jupiterone-dev.zendesk.com",
              "httpOnly": true,
              "name": "__cfruid",
              "path": "/",
              "sameSite": "None",
              "secure": true,
              "value": "[REDACTED]"
            }
          ],
          "headers": [
            {
              "name": "date",
              "value": "Fri, 28 Jan 2022 14:53:11 GMT"
            },
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "zendesk-api-version",
              "value": "2022-01-01"
            },
            {
              "name": "x-zendesk-api-version",
              "value": "v2"
            },
            {
              "name": "x-zendesk-application-version",
              "value": "v10911"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "etag",
              "value": "W/\"aee7c5535e2c5f8d50b0f209ef410549\""
            },
            {
              "name": "access-control-allow-origin",
              "value": "*"
            },
            {
              "name": "access-control-allow-credentials",
              "value": "true"
            },
            {
              "name": "access-control-expose-headers",
              "value": "X-Zendesk-API-Warn,X-Zendesk-User-Id,X-Zendesk-User-Session-Expires-At"
            },
            {
              "name": "cache-control",
              "value": "max-age=30, private"
            },
            {
              "name": "x-rate-limit",
              "value": "200"
            },
            {
              "name": "x-rate-limit-remaining",
              "value": "191"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000;"
            },
            {
              "name": "x-zendesk-origin-server",
              "value": "classic-app-server-7f77698bc-hc6mc"
            },
            {
              "name": "x-request-id",
              "value": "6d4b0c7e89c62a36-ORD, 6d4b0c7e89c62a36-ORD"
            },
            {
              "name": "x-runtime",
              "value": "0.151461"
            },
            {
              "name": "x-zendesk-zorg",
              "value": "yes"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "cf-cache-status",
              "value": "DYNAMIC"
            },
            {
              "name": "expect-ct",
              "value": "max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\""
            },
            {
              "name": "report-to",
              "value": "{\"endpoints\":[{\"url\":\"https:\\/\\/a.nel.cloudflare.com\\/report\\/v3?s=t%2BFJAiW%2BcmP961qhs9nc%2FEAZbBYmwNjo%2F%2F4p9qZIWr4QM0PKZQHJJ4DIWMiEBIXFhTgrIFCZmjxeKQYw6MTWCXmumeU6wtnZQWp%2FaguzyRnfT%2Bmo8qBwpsU5g0QpddeWDQJCG7Xoose3Z2q5\"}],\"group\":\"cf-nel\",\"max_age\":604800}"
            },
            {
              "name": "nel",
              "value": "{\"success_fraction\":0.01,\"report_to\":\"cf-nel\",\"max_age\":604800}"
            },
            {
              "_fromType": "array",
              "name": "set-cookie",
              "value": "[REDACTED]"
            },
            {
              "name": "server",
              "value": "cloudflare"
            },
            {
              "name": "cf-ray",
              "value": "6d4b0c7e89c62a36-ORD"
            }
          ],
          "headersSize": 1501,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-01-28T14:53:10.631Z",
        "time": 546,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 546
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
