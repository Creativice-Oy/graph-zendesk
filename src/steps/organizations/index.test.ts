import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';
import { buildOrganizationGroupRelationships, fetchOrganizations } from '.';
import { integrationConfig } from '../../../test/config';
import { setupZendeskRecording } from '../../../test/recording';
import { IntegrationConfig } from '../../config';
import { fetchAccount } from '../account';
import { Entities, Relationships } from '../constants';
import { fetchGroups } from '../groups';

describe('#fetchOrganizations', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupZendeskRecording({
      directory: __dirname,
      name: 'fetchOrganizations',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchAccount(context);
    await fetchOrganizations(context);

    const organizations = context.jobState.collectedEntities.filter((e) =>
      e._type.includes(Entities.ORGANIZATION._type),
    );

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    expect(organizations.length).toBeGreaterThan(0);
    expect(organizations).toMatchGraphObjectSchema({
      _class: Entities.ORGANIZATION._class,
      schema: {
        additionalProperties: false,
        required: [],
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: Entities.ORGANIZATION._type },
          _class: { const: Entities.ORGANIZATION._class },
          _key: { type: 'string' },
          createdAt: { type: 'string' },
          details: { type: 'array' },
          domainNames: { type: 'array' },
          groupId: { type: 'string' },
          id: { type: 'string' },
          name: { type: 'string' },
          notes: { type: 'array' },
          sharedComments: { type: 'boolean' },
          sharedTickets: { type: 'boolean' },
          updatedAt: { type: 'string' },
          url: { type: 'string' },
        },
      },
    });
  });
});

describe('#buildOrganizationGroupRelationships', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupZendeskRecording({
      directory: __dirname,
      name: 'buildOrganizationGroupRelationships',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchGroups(context);
    await fetchOrganizations(context);
    await buildOrganizationGroupRelationships(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    expect(
      context.jobState.collectedRelationships.filter(
        (e) => e._type === Relationships.ORGANIZATION_HAS_GROUP._type,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'HAS' },
          _type: {
            const: 'zendesk_organization_has_group',
          },
        },
      },
    });
  });
});
