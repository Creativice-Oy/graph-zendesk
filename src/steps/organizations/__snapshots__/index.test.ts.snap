// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#buildOrganizationGroupRelationships should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "zendesk_group:360021738231",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "created_at": "2021-10-10T20:49:30Z",
            "default": true,
            "deleted": false,
            "description": "",
            "id": 360021738231,
            "name": "Support",
            "updated_at": "2021-10-10T20:49:30Z",
            "url": "https://jupiterone-dev.zendesk.com/api/v2/groups/360021738231.json",
          },
        },
      ],
      "_type": "zendesk_group",
      "createdAt": "2021-10-10T20:49:30Z",
      "createdOn": undefined,
      "deleted": false,
      "description": "",
      "displayName": "Support",
      "id": "360021738231",
      "name": "Support",
      "updatedAt": "2021-10-10T20:49:30Z",
      "url": "https://jupiterone-dev.zendesk.com/api/v2/groups/360021738231.json",
    },
    Object {
      "_class": Array [
        "Organization",
      ],
      "_key": "zendesk_organization:361584776731",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "created_at": "2021-10-10T20:49:30Z",
            "details": "",
            "domain_names": Array [],
            "external_id": null,
            "group_id": 360021738231,
            "id": 361584776731,
            "name": "JupiterOne",
            "notes": "",
            "organization_fields": Object {},
            "shared_comments": false,
            "shared_tickets": false,
            "tags": Array [],
            "updated_at": "2021-11-05T16:50:42Z",
            "url": "https://jupiterone-dev.zendesk.com/api/v2/organizations/361584776731.json",
          },
        },
      ],
      "_type": "zendesk_organization",
      "createdAt": "2021-10-10T20:49:30Z",
      "createdOn": undefined,
      "details": Array [],
      "displayName": "JupiterOne",
      "domainNames": Array [],
      "groupId": "360021738231",
      "id": "361584776731",
      "name": "JupiterOne",
      "notes": Array [],
      "sharedComments": false,
      "sharedTickets": false,
      "updatedAt": "2021-11-05T16:50:42Z",
      "url": "https://jupiterone-dev.zendesk.com/api/v2/organizations/361584776731.json",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "zendesk_organization:361584776731",
      "_key": "zendesk_organization:361584776731|has|zendesk_group:360021738231",
      "_toEntityKey": "zendesk_group:360021738231",
      "_type": "zendesk_organization_has_group",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "zendesk_group",
    "zendesk_organization",
    "zendesk_organization_has_group",
  ],
  "numCollectedEntities": 2,
  "numCollectedRelationships": 1,
}
`;

exports[`#fetchOrganizations should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "zendesk_account:jupiterone-dev",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "multiproduct": true,
            "name": "JupiterOne",
            "owner_id": 420755677092,
            "sandbox": false,
            "subdomain": "jupiterone-dev",
            "time_format": 24,
            "time_zone": "America/New_York",
            "url": "https://jupiterone-dev.zendesk.com",
          },
        },
      ],
      "_type": "zendesk_account",
      "accessURL": "https://jupiterone-dev.zendesk.com",
      "accountId": "jupiterone-dev",
      "createdOn": undefined,
      "displayName": "JupiterOne",
      "name": "JupiterOne",
    },
    Object {
      "_class": Array [
        "Organization",
      ],
      "_key": "zendesk_organization:361584776731",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "created_at": "2021-10-10T20:49:30Z",
            "details": "",
            "domain_names": Array [],
            "external_id": null,
            "group_id": 360021738231,
            "id": 361584776731,
            "name": "JupiterOne",
            "notes": "",
            "organization_fields": Object {},
            "shared_comments": false,
            "shared_tickets": false,
            "tags": Array [],
            "updated_at": "2021-11-05T16:50:42Z",
            "url": "https://jupiterone-dev.zendesk.com/api/v2/organizations/361584776731.json",
          },
        },
      ],
      "_type": "zendesk_organization",
      "createdAt": "2021-10-10T20:49:30Z",
      "createdOn": undefined,
      "details": Array [],
      "displayName": "JupiterOne",
      "domainNames": Array [],
      "groupId": "360021738231",
      "id": "361584776731",
      "name": "JupiterOne",
      "notes": Array [],
      "sharedComments": false,
      "sharedTickets": false,
      "updatedAt": "2021-11-05T16:50:42Z",
      "url": "https://jupiterone-dev.zendesk.com/api/v2/organizations/361584776731.json",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "zendesk_account:jupiterone-dev",
      "_key": "zendesk_account:jupiterone-dev|has|zendesk_organization:361584776731",
      "_toEntityKey": "zendesk_organization:361584776731",
      "_type": "zendesk_account_has_organization",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "zendesk_account",
    "zendesk_organization",
    "zendesk_account_has_organization",
  ],
  "numCollectedEntities": 2,
  "numCollectedRelationships": 1,
}
`;
